### ISSUE TITLE
Add Megatron-LM precision options

### ISSUE THREAD

**Issue by kolehma8** [verl-project/verl#6576]
### Feature request

Hi Team,

Currently verl Megatron backend holds the model, gradients and optimizer states in FP32 precision regardless of the compute precision. Megatron-LM allows changing the gradient and/or optimizer state precisions to BF16 via flags that are not currently incorporated in verl. This can be highly desirable for use cases that can do away with lower precision while wanting to minimize the memory footprint of the model.

I recently created a PR (https://github.com/verl-project/verl/pull/6526) to address the above issue but it got closed without any follow up discussion. I wanted to ask if the Verl team could address this if OS contributions are not welcomed?

### Motivation

Allow reducing the memory footprint of the Megatron backend.

### Your contribution

https://github.com/verl-project/verl/pull/6526

**Comment by yueyiming2009** [comment]
Supporting BF16 for gradients and optimizer states in the Megatron backend is a highly valuable feature. When dealing with large-scale model training, minimizing the memory footprint is critical, and having the flexibility to adjust these compute precision flags is definitely needed for optimization.

**Comment by kolehma8** [comment]
# BF16 Optimizer-State / Gradient Experiments — Results

**Backend:** verl Megatron · **Algorithm:** GRPO · **Model:** Qwen2.5-7B-Instruct · **Dataset:** gsm8k
**Hardware:** 1× node, 8× H100 80GB
**Per run:** TP=4, PP=1 → DP=2; vLLM rollout TP=2; `n=5`; `total_epochs=15` (~105 training steps)
**Design:** 4 precision configs × 5 seeds = **20 runs** (all completed, 0 failures). The model compute stays BF16 (standard mixed precision) in every config; only the **optimizer-state** and **gradient-buffer** dtypes vary.

## Precision configurations

| config | optimizer states (`exp_avg`, `exp_avg_sq`) | gradients (`main_grads_dtype` + DDP all-reduce) |
|--------|--------------------------------------------|--------------------------------------------------|
| `baseline`  | FP32 | FP32 |
| `opt_bf16`  | **BF16** | FP32 |
| `grad_bf16` | FP32 | **BF16** |
| `all_bf16`  | **BF16** | **BF16** |

## Training reward (`critic/score/mean`), mean over 5 seeds

| config | final-step mean ± std | last-10-step mean ± std |
|--------|------------------------|--------------------------|
| `baseline`  | 0.9777 ± 0.0026 | 0.9798 ± 0.0011 |
| `opt_bf16`  | 0.9781 ± 0.0021 | 0.9795 ± 0.0012 |
| `grad_bf16` | 0.9778 ± 0.0032 | 0.9805 ± 0.0011 |
| `all_bf16`  | 0.9775 ± 0.0032 | 0.9798 ± 0.0011 |

All four configs converge to **~0.978** gsm8k reward. The largest gap between any config and the FP32 baseline is **≤ 0.0006** (final-step) — roughly **5× smaller** than the seed-to-seed standard deviation (~0.002–0.003). **The differences are statistical noise.**

## Peak GPU memory (`perf/max_memory_allocated_gb`, max across steps/seeds)

| config | peak allocated | peak reserved | savings vs baseline |
|--------|----------------|---------------|----------------------|
| `baseline`  | 47.87 GB | 49.18 GB | — |
| `opt_bf16`  | 42.36 GB | 44.06 GB | −5.5 GB |
| `grad_bf16` | 43.12 GB | 44.46 GB | −4.8 GB |
| `all_bf16`  | **39.37 GB** | 41.02 GB | **−8.5 GB (~18%)** |

## Verdict

> **Using BF16 for the optimizer states and/or gradients matches FP32 training reward** on gsm8k + Qwen2.5-7B GRPO over 105 steps × 5 seeds, while reducing peak GPU memory by up to **~18%** in the full `all_bf16` case. The memory savings come essentially for free — no measurable reward degradation, individually or combined.

Notes:
- DP=2 (TP=4 on 8 GPUs), so the BF16 **gradient all-reduce** across data-parallel ranks *is* exercised in `grad_bf16`/`all_bf16`, along with the BF16 grad-accumulation buffer (8 micro-batches/optimizer step) and BF16 Adam moments.
- OOM under colocation was avoided with `rollout.gpu_memory_utilization=0.3` (gsm8k sequences are short, ~254 tokens, so a small vLLM KV pool suffices) + gradient checkpointing. TP=8 is **not** valid for this model (`num_attention_heads=28` is not divisible by 8).

**Comment by kolehma8** [comment]
Reward curves for the training with the first seed for every BF16 and baseline ablation:
<img width="993" height="459" alt="Image" src="https://github.com/user-attachments/assets/9e38dbdd-91b4-46a2-b5c9-64a78488bd5d" />

**Comment by Mecoli1219** [comment]
+1 on this. Hit the same wall on the Megatron backend.

**Comment by kolehma8** [comment]
PR (https://github.com/verl-project/verl/pull/6526) merged.
